module/modmanager/manager.go

```diff
@@ -0,0 +1,370 @@
+	if err != nil {
+		return err
+	}
+	_, err = module.client.ReconfigureComponent(ctx, &pb.ReconfigureComponentRequest{Config: cfgProto, Dependencies: deps})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsModularService returns true if a service would be handled by a module.
+func (mgr *Manager) IsModularService(cfg config.Service) bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	return mgr.getServiceModule(cfg) != nil
+}
+
+// ReconfigureService updates/reconfigures a modular service with a new configuration.
+func (mgr *Manager) ReconfigureService(ctx context.Context, cfg config.Service) error {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	module := mgr.getServiceModule(cfg)
+	if module == nil {
+		return errors.Errorf("no module registered to serve resource api %s and model %s", cfg.ResourceName().Subtype, cfg.Model)
+	}
+
+	cfgProto, err := config.ServiceConfigToProto(&cfg)
+	if err != nil {
+		return err
+	}
+	_, err = module.client.ReconfigureService(ctx, &pb.ReconfigureServiceRequest{Config: cfgProto})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsModularResource returns true if a component is handled by a module.
+func (mgr *Manager) IsModularResource(name resource.Name) bool {
+	mgr.mu.RLock()
+	defer mgr.mu.RUnlock()
+	_, ok := mgr.rMap[name]
+	return ok
+}
+
+// RemoveResource requests the removal of a resource from a module.
+func (mgr *Manager) RemoveResource(ctx context.Context, name resource.Name) error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	module, ok := mgr.rMap[name]
+	if !ok {
+		return errors.Errorf("resource %+v not found in module", name)
+	}
+
+	_, err := module.client.RemoveResource(ctx, &pb.RemoveResourceRequest{Name: name.String()})
+	return err
+}
+
+func (mgr *Manager) getComponentModule(cfg config.Component) *module {
+	for _, module := range mgr.modules {
+		var api resource.RPCSubtype
+		var ok bool
+		for a := range module.handles {
+			if a.Subtype == cfg.ResourceName().Subtype {
+				api = a
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			continue
+		}
+		for _, model := range module.handles[api] {
+			if cfg.Model == model {
+				return module
+			}
+		}
+	}
+	return nil
+}
+
+func (mgr *Manager) getServiceModule(cfg config.Service) *module {
+	for _, module := range mgr.modules {
+		var api resource.RPCSubtype
+		var ok bool
+		for a := range module.handles {
+			if a.Subtype == cfg.ResourceName().Subtype {
+				api = a
+				ok = true
+				break
+			}
+		}
+		if !ok {
+			continue
+		}
+		for _, model := range module.handles[api] {
+			if cfg.Model == model {
+				return module
+			}
+		}
+	}
+	return nil
+}
+
+// DepsToNames converts a dependency list to a simple string slice.
+func DepsToNames(deps registry.Dependencies) []string {
+	var depStrings []string
+	for dep := range deps {
+		depStrings = append(depStrings, dep.String())
+	}
+	return depStrings
+}
+
+func (m *module) checkReady(ctx context.Context, addr string) error {
```

Comment:
should we flip this to be that the client will call out to the server that it's ready?

Reply:
That was my original thought during early brainstorming, but... I believe this is simpler as is for a couple reasons:

1. The code flow is simpler if it can just wait for the module to start in the same function that's loading the module, rather than having to setup a channel to catch the incoming "ready" from another thread that's answering an incoming grpc request.
2. Currently, only one new proto was needed, and this represents the Module side, so no need for a second proto for the parent.
3. That also means less entanglement between the WebService and the module system. The "Lite" part of the WebService only handles resource services, same as the main web service does, so very little had to be changed there.
4. Lastly, this often means half as many sockets, as only modules that NEED a parent component ever connect to the parent. Having the module "check in" would mean there'd always be two connections per module.

Reply:
okay fair points!